README.md
# rcrtan

[![Build Status](https://travis-ci.org/gtlaflair/rcrtan.svg?branch=master)](https://travis-ci.org/gtlaflair/rcrtan) ![CRAN](http://www.r-pkg.org/badges/version/rcrtan)

rcrtan provides functions for criterion-referenced test analyses as described in Brown & Hudson (2002). Currently it supports the following item and test analyses:

* Item discrimination
     + Item facility (including IF for passing test takers and failing test takers)
     + B-index
     + Agreement index
     + Item Phi
  
* Test dependability
     + Subkoviak's (1988) single administration kappa coefficient
     + Subkoviak's (1988) single administration agreement coefficient
     + Brown's (1990) short-cut estimate for phi dependability
     + Brennan's (1984) estimate for phi lambda

# Installation & Documentation

Install the most recent version from CRAN

```R
install.packages('rcrtan')
```

Or, get the development version from github:

```R
install.packages('devtools')
devtools::install_github('gtlaflair/rcrtan')
```
# Item analysis functions

This package contains six unique functions for calculating item statistics and one omnibus function for estimating all of the six unique functions. The formulas have been adapted from Brown and Hudson (2002).

* `crt_iteman`: omnibus function that estimates item facility, item facility for masters, item facility for non-masters, the B-index, the agreement statistic, and item phi.

* `if_total`: estimates the item facility values for the entire group of test takers.

<!-- _k_ is the total number of dichotomously score items on a test.  -->

<!-- $$\frac{\sum_{i=1}^k X_{i}}{k}$$ -->

* `if_pass`: estimates the item facility values for the test takers who scored above the cut-score (masters).

<!-- $$if\_pass = \frac{\sum_{i=1}^k X_{i_{masters}}}{k}$$ -->

* `if_fail`: estimates the item facility values for the test takers who scored below the cut-score (non-masters).

<!-- $$if\_fail = \frac{\sum_{i=1}^k X_{i_{non-masters}}}{k}$$ -->

* `b_index`: estimates the B-index values for each item. This is the difference in item facility for masters and non-masters.

<!-- $$b\_index = if\_pass - if\_fail$$ -->

* `agree_stat`: estimates the agreement statistic. 

<!-- _P_~iT~ is the proportion ot test takers who passed the test and answered the item correctly. _Q_~i~ is the propotion of test takers who answered the item incorrectly. _P_~T~ is the proportion of test takers who passed the test. -->

<!-- $$agree\_stat = 2 * P_{iT} + Q_{i} - P_{T}$$ -->

* `item_phi`: estimates item phi. 

<!-- _P_~iT~ is the proportion ot test takers who passed the test and answered the item correctly. _P_~i~ is the proportion of test takers who answered the item correctly. _P_~T~ is the proportion of test takers who passed the test. _Q_~i~ is the proportion of test takers who answered the item incorrectly. _Q_~T~ is the proportion of examinees who failed the test. -->

<!-- $$item\_phi = P_{iT} - P_{i} * P_{T} /  \sqrt{P_{i} * Q_{i} * P_{T} * Q_{T}}$$ -->

```{r}
item_stats <- crt_iteman(data = brown_depend, items = 2:31, raw_cut_score = 21, scale = 'raw')
```

# Test dependability functions

This package contains two functions for estimating test dependability. The formulas have been adapted from Brown and Hudson (2002).

* `subkoviak`: This function uses Subkoviak's (1988) methods for estimating kappa and rho from a single test administration.

<!-- _z_ is the standardized cut-score. _c_ is the raw cut-score. _M_ is the mean of the raw test scores. _S_ is the standard deviation of the raw test scores, and 0.5 is a constant adjustment factor. This function uses the _z_ cut-score and an estimate of coefficient reliability, both rounded to the nearest $\frac{1}{10}$, in order to look up the single-test administration $\kappa$ and $\rho$ estimates. -->

<!-- $$ z = \frac{c - 0.5 - M}{S} $$ -->

```{r}
sub_estimate <- subkoviak(data = brown_depend, items = 2:31, raw_cut_score = 21, look_up = FALSE)

```

* `phi_domain`: This function uses Brown's (1990) formula for a short-cut estimate of generalizability coefficients for absolute decisions. 

<!-- _n_ is the number of test takers. _k_ is the number of items on the test. _M_~p~ is the mean of the total scores divided by the number of items. _S_~p~^2^ is the standard deviation of the total scores divided by the number of items squared. $\alpha$ is the coefficient alpha reliability estimate. -->

<!-- $$ \phi = \frac{\frac{n * S_{p}^2}{n - 1} * \alpha}{\frac{n * S_{p}^2}{n - 1} * \alpha + \frac{M_{p} * (1 - M_{p}) - S_{p}^2}{k-1}}$$ -->

```{r}
phi_estimate <- phi_domain(bh_depend, 2:31)
```


## References

Brennan, R.L. (1984). Estimating the dependability of the scores. In R.A. Berk (Ed.) _A guide to criterion-referenced test construction_ (pp. 292-334). Baltimore: Johns Hopkins University Press.

Brown, J.D. (1990). Short-cut estimates for criterion-referenced test consistency. _Language Testing, 7_(1), 77-97.

Brown, J.D., & Hudson, T (2002). _Criterion-referenced language testing_. Cambridge: Cambridge University Press.

Subkoviak, M.J. (1988). A practitioner's guide to computation and interpretation of reliability indices for mastery tests. _Journal of Educational Measurement, 25_, 47-55.

NEWS.md
# rcrtan 0.1.1

## Major changes

- none

## Bug fixes

- The functions have been updated to avoid bugs related to the release of `tibble` 2.0

R/item_desc.R

#' Calculate item facility
#' 
#' @importFrom magrittr %>%
#' @importFrom dplyr summarise_all
#' @importFrom tibble as_tibble
#' @importFrom dplyr rename
#' 
#' @export if_total
#' 
#' @param data A data frame of dichotomously scored test times
#' @param items Raw column indices representing the test items
#' @return Item_facility Item facility values for test items
#' 
#' @examples 
#' if_total(bh_depend, 2:31)

if_total <- function(data, items){
  
  Item_facility <- data %>%
    select(., items) %>%
    summarise_all(funs(mean)) %>%
    t(.)%>%
    as.data.frame(.) %>%
    as_tibble(., rownames = 'items') %>%
    rename(., 'if_total' = "V1")
  
  return(Item_facility)
  
}

#' Calculate item facility for passing students
#' 
#' @importFrom stats setNames
#' @importFrom magrittr %>% 
#' @importFrom dplyr filter
#' @importFrom dplyr summarise_all
#' @importFrom tibble as_tibble
#' @importFrom dplyr rename
#' 
#' @export if_pass
#' 
#' @param data A data frame of dichotomously scored test times
#' @param items Raw column indices representing the test items
#' @param cut_score A raw or percentage cut-score
#' @param scale A character vector indicating whether the cut-score
#' is 'raw' (default) or 'percent'
#' @return Item_facility_pass Item facility values for test items of
#' of test takers who passed the test
#' 
#' @examples 
#' if_pass(bh_depend, 2:31, 21, scale = 'raw')

if_pass <- function(data, items, cut_score, scale = 'raw'){
  
  data$raw_total <- data %>%
    select(., items) %>%
    rowSums(.)
  
  data$perc_total <- data %>%
    select(., items) %>% {
      (rowSums(.) / length(items)) * 100
    }
  
  if (scale == 'percent') {
    data$total <- data$perc_total
  }
  
  if (scale == 'raw') {
    data$total <- data$raw_total
  }
  
  
  data$pass <- ifelse(data$total >= cut_score, 'pass', 'fail')
  
  Item_facility_pass <- data %>%
    filter(pass %in% 'pass') %>%
    select(., items) %>%
    summarise_all(funs(mean)) %>%
    t(.) %>%
    as.data.frame(.) %>%
    as_tibble(., rownames = 'items') %>%
    rename(., 'if_pass' = "V1")
    

  return(Item_facility_pass)
}

#' Calculate item facility for failing students
#' 
#' @importFrom magrittr %>%
#' @importFrom dplyr filter
#' @importFrom dplyr summarise_all
#' @importFrom tibble as_tibble
#' @importFrom dplyr rename
#' 
#' @export if_fail
#' 
#' @param data A data frame of dichotomously scored test times
#' @param items Raw column indices representing the test items
#' @param cut_score A raw or percentage cut-score
#' @param scale A character vector indicating whether the cut-score
#' is 'raw' (default) or 'percent'
#' @return Item_facility_fail Item facility values for test items of
#' of test takers who failed the test
#' 
#' @examples 
#' if_fail(bh_depend, 2:31, 21, scale = 'raw')

if_fail <- function(data, items, cut_score, scale = 'raw'){
  
  data$raw_total <- data %>%
    select(., items) %>%
    rowSums(.)
  
  data$perc_total <- data %>%
    select(., items) %>% {
      (rowSums(.) / length(items)) * 100
    }
  
  if (scale == 'percent') {
    data$total <- data$perc_total
  }
  
  if (scale == 'raw') {
    data$total <- data$raw_total
  }
  
  data$pass <- ifelse(data$total >= cut_score, 'pass', 'fail')
  
  Item_facility_fail <- data %>%
    filter(pass %in% 'fail') %>%
    select(., items) %>%
    summarise_all(funs(mean)) %>%
    t(.) %>%
    as.data.frame(.) %>%
    as_tibble(., rownames = 'items') %>%
    rename(., 'if_fail' = "V1")
  
  return(Item_facility_fail)
}

#' Calculate B-index
#' 
#' @importFrom magrittr %>%
#' @importFrom dplyr filter
#' @importFrom tibble as_data_frame
#' @importFrom dplyr summarise_all
#' @importFrom tibble as_tibble
#' @importFrom dplyr rename
#' 
#' @export b_index
#' 
#' @param data A data frame of dichotomously scored test times
#' @param items Raw column indices representing the test items
#' @param cut_score A raw or percentage cut-score
#' @param scale A character vector indicating whether the cut-score
#' is 'raw' (default) or 'percent'
#' @return Bindex B-index values for items on the test
#' 
#' @examples 
#' b_index(bh_depend, 2:31, 21, scale = 'raw')

b_index <- function(data, items, cut_score, scale = 'raw'){
  
  data$raw_total <- rowSums(data[items])
  data$perc_total <- (rowSums(data[items]) / length(items)) * 100
  
  ifelse(scale == 'percent', data$total <- data$perc_total, data$total <- data$raw_total)
  
  data$pass <- ifelse(data$total >= cut_score, 'pass', 'fail')
  
  Item_facility_pass <- data %>%
    filter(pass %in% 'pass') %>%
    select(., items) %>%
    summarise_all(., funs(mean)) 
    
  Item_facility_fail <- data %>%
    filter(pass %in% 'fail') %>%
    select(., items) %>%
    summarise_all(., funs(mean)) 
    
  Bindex <- Item_facility_pass - Item_facility_fail
  
  Bindex <- Bindex %>%
    t(.) %>%
    as.data.frame(.) %>%
    as_tibble(., rownames = 'items') %>%
    rename(., 'b_index' = "V1")
  
  return(Bindex)
}

#' Calculate Agreement statistic
#' 
#' @importFrom magrittr %>% 
#' @importFrom dplyr filter
#' @importFrom dplyr summarise_all
#' @importFrom dplyr summarise
#' @importFrom dplyr funs
#' @importFrom tibble as_tibble
#' @importFrom dplyr rename
#' 
#' @export agree_stat
#' 
#' @param data A data frame of dichotomously scored test times
#' @param items Raw column indices representing the test items
#' @param cut_score A raw or percentage cut-score
#' @param scale A character vector indicating whether the cut-score
#' is 'raw' (default) or 'percent'
#' @return Agree Agreement statistic values for items on the test
#' 
#' @examples 
#' agree_stat(bh_depend, 2:31, 21, scale = 'raw')

agree_stat <- function(data, items, cut_score, scale = 'raw'){
  
  data$raw_total <- rowSums(data[items])
  data$perc_total <- (rowSums(data[items]) / length(items)) * 100
  
  ifelse(scale == 'percent', data$total <- data$perc_total, data$total <- data$raw_total)
  
  data$pass <- ifelse(data$total >= cut_score, 'pass', 'fail')

  PiT <- data %>%
    filter(pass %in% 'pass') %>%
    select(., items) %>%
    summarise_all(funs(sum)) %>%
    .[] / length(data[[1]])

  Qi <- data %>%
    select(., items) %>%
    summarise_all(funs(counts=sum(. == 0,na.rm=TRUE))) %>%
    .[] / length(data[[1]])

  Pt <- data %>%
    summarise(pass_prop = length(which(pass %in% 'pass'))/length(pass)) %>%
    rep(.,length(items)) %>%
    unlist

  Agree <- (2 * PiT) + Qi - Pt 
  
  Agree <- Agree %>%
    t(.) %>%
    as.data.frame(.) %>%
    as_tibble(., rownames = 'items') %>%
    rename(., 'agree' = "V1")

  return(Agree)
}

#' Calculate Item Phi
#' 
#' @importFrom magrittr %>%
#' @importFrom dplyr filter
#' @importFrom dplyr summarise_all
#' @importFrom dplyr funs
#' @importFrom dplyr rename
#' 
#' @export item_phi
#' 
#' @param data A data frame of dichotomously scored test times
#' @param items Raw column indices representing the test items
#' @param cut_score A raw or percentage cut-score
#' @param scale A character vector indicating whether the cut-score
#' is 'raw' (default) or 'percent'
#' @return Phi Item Phi values for items on the test
#' 
#' @examples 
#' item_phi(bh_depend, 2:31, 21, scale = 'raw')

item_phi <- function(data, items, cut_score, scale = 'raw'){
  
  data$raw_total <- rowSums(data[items])
  data$perc_total <- (rowSums(data[items]) / length(items)) * 100
  
  ifelse(scale == 'percent', data$total <- data$perc_total, data$total <- data$raw_total)
  
  data$pass <- ifelse(data$total >= cut_score, 'pass', 'fail')
  
  PiT <- data %>%
    filter(pass %in% 'pass') %>%
    select(., items) %>%
    summarise_all(funs(sum)) %>%
    .[] / length(data[[1]])
  
  Qi <- data %>%
    select(., items) %>%
    summarise_all(funs(counts=sum(. == 0,na.rm=TRUE))) %>%
    .[]/length(data[[1]])
  
  Pi <- 1 - Qi
  
  Pt <- data %>%
    summarise(pass_prop = length(which(pass %in% 'pass'))/length(pass)) %>%
    rep(.,length(items)) %>%
    unlist
  
  Qt <- 1 - Pt
  
  Phi <- (PiT - (Pi*Pt)) / (sqrt(Pi * Qi * Pt * Qt))
  
  Phi <- Phi %>%
    t(.) %>%
    as.data.frame(.) %>%
    as_tibble(., rownames = 'items') %>%
    rename(., 'phi' = "V1")
  
  return(Phi)
  
}

#' Calculate criterion-referenced item discrimination indices
#' 
#' @importFrom stats setNames
#' @importFrom magrittr %>%
#' @importFrom tibble as_data_frame
#' @importFrom dplyr full_join
#' 
#' @export crt_iteman
#' 
#' @param data A data frame of dichotomously scored test times
#' @param items Raw column indices representing the test items
#' @param cut_score A raw or percentage cut-score
#' @param scale A character vector indicating whether the cut-score
#' is 'raw' (default) or 'percent'
#' @return \code{if_pass} contains item facility values for test items for students who 
#' passed the test
#' @return \code{if_fail} contains item facility values for test items for students who did not pass the test
#' @return \code{if_total} contains item facility values for test items 
#' @return \code{b_index} contains b-index values for items on the test
#' @return \code{agree_stat} contains agreement statistic values for items on the test
#' @return \code{item_phi} contains item phi values for items on the test
#' 
#' @examples 
#' crt_iteman(bh_depend, 2:31, 21, scale = 'raw')

crt_iteman <- function(data, items, cut_score, scale = 'raw'){

  iteman <- data %>% {
    
    item_fac <- if_total(., items = items) 

    item_fac_pass <- if_pass(., items = items, cut_score = cut_score, scale = scale) 
    
    item_fac_fail <- if_fail(., items = items, cut_score = cut_score, scale = scale) 
    
    b <- b_index(.,items = items, cut_score = cut_score, scale = scale) 
    
    a <- agree_stat(.,items = items, cut_score = cut_score, scale = scale) 
    
    p <- item_phi(.,items = items, cut_score = cut_score, scale = scale) 
    
    res <- full_join(item_fac_pass, item_fac_fail, by = "items") %>%
      full_join(., item_fac, by = "items") %>%
      full_join(., b, by = "items") %>%
      full_join(., a, by = "items") %>%
      full_join(., p, by = "items")
    
    return(res)
  }
  
  return(iteman)
}

globalVariables(c('pass', '.', 'total'))